### PR TITLE
Allow conversion of collection initializer Add extension method this arg

### LINKED ISSUE #38726: Incorrect IL generated for struct collection initializer call to Add extension method
Compile:
```C#
using System;
using System.Collections;

interface IAppend
{
    void Append(object o);
}

struct S : IEnumerable, IAppend
{
    IEnumerator IEnumerable.GetEnumerator() => null;
    void IAppend.Append(object o) { }
}

static class Program
{
    static void Add(this IAppend x, object y)
    {
        x.Append(y);
    }
    static void Main()
    {
        _ = new S() { 1, 2 };
    }
}
```
Execute:
```
Unhandled Exception: System.NullReferenceException:
    Object reference not set to an instance of an object.
   at Program.Add(IAppend x, Object y)
   at Program.Main()
```
See https://github.com/dotnet/roslyn/issues/30956#issuecomment-435804724 for related issue with `await new S()`.

### PR DESCRIPTION
**Pull request by cston**
Fixes #38726

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 1]
```diff
@@ -1640,6 +1643,7 @@
   </Node>
 
   <Node Name="BoundCollectionInitializerExpression" Base="BoundObjectInitializerExpressionBase">
+    <Field Name="PlaceholderOpt" Type="BoundLValuePlaceholder" Null="allow" />
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >BoundLValuePlaceholder [](start = 39, length = 22)

The name feels somewhat confusing. Usually, LValue is used to refer to something that can be assigned to, a local, a field, an array element, etc.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: ><Field Name="PlaceholderOpt" [](start = 4, length = 28)

Consider adding a comment describing how the placeholder is used.

**Comment by AlekseyTs** [comment]
>                         var implicitReceiver = new BoundImplicitReceiver(typeSyntax, type) { WasCompilerGenerated = true };

Is there a benefit to continue using BoundImplicitReceiver here?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4156 in 1b3ef89. [](commit_id = 1b3ef89a20b8b22a34337f15cf2fbbb23e4a6876, deletion_comment = False)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -1882,7 +1882,6 @@ private new void VisitCollectionElementInitializer(BoundCollectionElementInitial
             (var reinferredMethod, _, _) = VisitArguments(node, node.Arguments, refKindsOpt: default, node.AddMethod, node.ArgsToParamsOpt, node.Expanded, node.InvokedAsExtensionMethod);
             if (node.ImplicitReceiverOpt != null)
             {
-                Debug.Assert(node.ImplicitReceiverOpt.Kind == BoundKind.ImplicitReceiver);
                 SetAnalyzedNullability(node.ImplicitReceiverOpt, new VisitResult(node.ImplicitReceiverOpt.Type, NullableAnnotation.NotAnnotated, NullableFlowState.NotNull));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >SetAnalyzedNullability(node.ImplicitReceiverOpt, new VisitResult(node.ImplicitReceiverOpt.Type, NullableAnnotation.NotAnnotated, NullableFlowState.NotNull)); [](start = 16, length = 157)

Will it dig through conversion?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -7566,6 +7565,12 @@ public override BoundNode VisitDeconstructValuePlaceholder(BoundDeconstructValue
             return null;
         }
 
+        public override BoundNode VisitLValuePlaceholder(BoundLValuePlaceholder node)
+        {
+            SetUnknownResultNullability(node);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >SetUnknownResultNullability(node); [](start = 12, length = 34)

I guess unknown would work, but, given the usage, should we use SetNotNullResult instead?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: This is what we do in ```VisitImplicitReceiver``` 

---
In reply to: [325386703](https://github.com/dotnet/roslyn/pull/38732#discussion_r325386703) [](ancestors = 325386703)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 4]
```diff
@@ -111,6 +111,9 @@
   <Node Name="BoundDisposableValuePlaceholder" Base="BoundValuePlaceholderBase">
   </Node>
 
+  <Node Name="BoundLValuePlaceholder" Base="BoundValuePlaceholderBase">
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >BoundLValuePlaceholder [](start = 14, length = 22)

It looks like there is a special handling for BoundImplicitReceiver in CSharpOperationFactory. We are using this new node in place where BoundImplicitReceiver was used. It feels like adjustments should be made to CSharpOperationFactory.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs [hunk 5]
```diff
@@ -36,14 +37,17 @@ private static BoundObjectInitializerExpressionBase UpdateInitializers(BoundObje
...
-                case BoundKind.ObjectInitializerExpression:
-                    AddObjectInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenReceiver, ((BoundObjectInitializerExpression)initializerExpression).Initializers);
+                case BoundObjectInitializerExpression objectInitializer:
+                    AddObjectInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenReceiver, objectInitializer.Initializers);
                     return;
 
-                case BoundKind.CollectionInitializerExpression:
-                    AddCollectionInitializers(ref dynamicSiteInitializers, result, rewrittenReceiver, ((BoundCollectionInitializerExpression)initializerExpression).Initializers);
+                case BoundCollectionInitializerExpression collectionInitializer:
+                    var placeholder = collectionInitializer.PlaceholderOpt;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs]: >PlaceholderOpt [](start = 60, length = 14)

Can this be null?

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: This comment is for name of the type ```BoundLValuePlaceholder```

---
In reply to: [325362467](https://github.com/dotnet/roslyn/pull/38732#discussion_r325362467) [](ancestors = 325362467)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >Null="allow" [](start = 63, length = 12)

Do we ever store null in this property?

File: src/Compilers/CSharp/Test/Emit/CodeGen/ObjectAndCollectionInitializerTests.cs [hunk 6]
```diff
@@ -3409,6 +3405,79 @@ static int Index
...
+{
+    IEnumerator IEnumerable.GetEnumerator() => null;
+    void IAppend.Append(object o) { }
+}
+
+static class Program
+{
+    static void Add(this IAppend x, object y)
+    {
+        x.Append(y);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/ObjectAndCollectionInitializerTests.cs]: >x.Append(y); [](start = 8, length = 12)

Consider printing something to verify that the method is called.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs [hunk 7]
```diff
@@ -36,14 +37,17 @@ private static BoundObjectInitializerExpressionBase UpdateInitializers(BoundObje
...
+                    AddPlaceholderReplacement(placeholder, rewrittenReceiver);
+                    AddCollectionInitializers(ref dynamicSiteInitializers, result, rewrittenReceiver, collectionInitializer.Initializers);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs]: >rewrittenReceiver [](start = 83, length = 17)

Consider passing down the placeholder instead and relying on the placeholder replacement for all scenarios not just for extension methods

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs]: It is possible we don't even need passing the placeholder.

---
In reply to: [325394891](https://github.com/dotnet/roslyn/pull/38732#discussion_r325394891) [](ancestors = 325394891)

**Comment by AlekseyTs** [comment]
>                 case BoundImplicitReceiver implicitReceiver:

Do we need to handle placeholder here?

---
Refers to: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:658 in e744041. [](commit_id = e744041a19767b03cbaabbd79879d2e3f0fe3693, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 8]
```diff
@@ -2092,5 +2095,15 @@ internal IOperation CreatePropertySubpatternMember(Symbol symbol, ITypeSymbol ma
                     return OperationFactory.CreateInvalidOperation(_semanticModel, nameSyntax, ImmutableArray<IOperation>.Empty, isImplicit);
             }
         }
+
+        private IInstanceReferenceOperation CreateLValuePlaceholderOperation(BoundLValuePlaceholder placeholder)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >CreateLValuePlaceholderOperation [](start = 44, length = 32)

Do we have an IOperation test covering scenario with conversion?

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 4 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Consider a comment on why this is needed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 9]
```diff
@@ -4552,7 +4552,7 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
             InitializerExpressionSyntax initializerSyntax,
             TypeSymbol initializerType,
             DiagnosticBag diagnostics,
-            BoundImplicitReceiver implicitReceiver)
+            BoundLValuePlaceholder implicitReceiver)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: We are using the term "lvalue" here but the implicit receiver is never an lvalue, it's always an rvalue (at least from the code perspective).

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs [hunk 10]
```diff
@@ -16,13 +16,14 @@ internal sealed partial class LocalRewriter
     {
         private static BoundObjectInitializerExpressionBase UpdateInitializers(BoundObjectInitializerExpressionBase initializerExpression, ImmutableArray<BoundExpression> newInitializers)
         {
-            if (initializerExpression.Kind == BoundKind.ObjectInitializerExpression)
+            switch (initializerExpression)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs]: Nit: could use a switch expression

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 8 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I don't see a test added for this.

---
In reply to: [325436181](https://github.com/dotnet/roslyn/pull/38732#discussion_r325436181) [](ancestors = 325436181)

**Comment by 333fred** [comment]
Done review pass (commit 4)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs (hunk 10 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs]: Yes, I considered that but decided on a switch statement to match the method below.

---
In reply to: [325892063](https://github.com/dotnet/roslyn/pull/38732#discussion_r325892063) [](ancestors = 325892063)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 5). Are failures of CI builds expected?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: The receiver is always the collection instance before conversions. I've added back an assert.

---
In reply to: [325386023](https://github.com/dotnet/roslyn/pull/38732#discussion_r325386023) [](ancestors = 325386023)

**Comment by cston** [comment]
@AlekseyTs, @333fred, @jaredpar, feedback has been addressed.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs (hunk 7 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs]: There are a number of places where the rewritten receiver are used in the generated code and where there is no placeholder instance on the initial bound node. (See dynamic cases in `AddObjectInitializer` for instance.) We could pass the placeholder and look up the replacement at those locations but it seems easier to just pass the replacement directly.

---
In reply to: [325395424](https://github.com/dotnet/roslyn/pull/38732#discussion_r325395424) [](ancestors = 325395424,325394891)

**Pull request merged**